### PR TITLE
Add Data Prepper Overview

### PR DESCRIPTION
**Pull request by Naarcha-AWS**
Signed-off-by: Naarcha-AWS <naarcha@amazon.com>

### Description

Adds Data Prepper overview to main vs Data Prepper overview to the data-prepper-2.0 repo. Brings over content and feedback from Data Prepper 2.0 #1304.

Cannot merge before the release of Data Prepper 2.0.

### Issues Resolved

Part of issue #1267 

### Checklist
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: _clients/data-prepper/index.md [hunk 1]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+- (Optional) One [buffer](#buffer)
+- (Optional) One or more[processors](#processor)
+
+A single instance of Data Prepper can have one or more pipelines. 
+
+Each pipeline definition contains two required components **source** and **sink**. If buffers and processors are missing from the Data Prepper pipeline, Data Prepper uses the default buffer and a no-op processor. 
+
+### Source 
+
+Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or reading from external endpoints like OTeL Collector for traces and metrics and S3. Source have their own configuration options based on the format of the events (such as string, json,  cloudwatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component.
```
Comment (natebower) [inline _clients/data-prepper/index.md:30]:
```suggestion
Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or by reading from external endpoints like OTeL Collector for traces and metrics and Amazon Simple Storage Service (Amazon S3). Sources have their own configuration options based on the format of the events (such as string, JSON, Amazon CloudWatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component. 
```

Comment (natebower) [inline _clients/data-prepper/index.md:30]: Should the first sentence start with "The source component"? In the third sentence, should it read something like "...or by reading from external endpoints, like OTeL Collector for traces, metrics, and Amazon Simple Storage Service (Amazon S3)"?

File: _clients/data-prepper/index.md [hunk 2]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+
+Each pipeline definition contains two required components **source** and **sink**. If buffers and processors are missing from the Data Prepper pipeline, Data Prepper uses the default buffer and a no-op processor. 
+
+### Source 
+
+Source is the input component of a pipeline that defines the mechanism through which a Data Prepper pipeline will consume events. A pipeline can have only one source. The source can consume events either by receiving the events over HTTP or HTTPS or reading from external endpoints like OTeL Collector for traces and metrics and S3. Source have their own configuration options based on the format of the events (such as string, json,  cloudwatch logs, or open telemetry trace). The source component consumes events and writes them to the buffer component. 
+
+### Buffer
+
+The buffer component acts as the layer between the source and the sink. Buffer can be either in-memory or disk-based. The default buffer uses an in-memory queue bounded by the number of events, called `bounded_blocking`. If the buffer component is not explicitly mentioned in the pipeline configuration, Data Prepper uses the default `bounded_blocking`.
```
Comment (natebower) [inline _clients/data-prepper/index.md:34]: Second sentence: "The buffer..."?

Comment (natebower) [inline _clients/data-prepper/index.md:34]:
```suggestion
The buffer component acts as the layer between the source and the sink. Buffer can be either in-memory or disk based. The default buffer uses an in-memory queue called `bounded_blocking` that is bounded by the number of events. If the buffer component is not explicitly mentioned in the pipeline configuration, Data Prepper uses the default `bounded_blocking`.
```

File: _clients/data-prepper/index.md [hunk 3]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+
+### Sink
+
+Sink is the output component of a pipeline that defines the destination(s) to which a Data Prepper pipeline publishes events. A sink destination could be services such as OpenSearch, S3, or another Data Prepper pipeline. When using another Data Prepper pipeline as the sink, you can chain multiple pipelines together based on the needs to the data. Sink contains it's own configurations options based on the destination type.
```
Comment (natebower) [inline _clients/data-prepper/index.md:38]: Should the first sentence start with "The sink component..."? Same in the third sentence: "When using another Data Prepper pipeline as the sink component..."? And in the fourth sentence: "The sink component contains..."? Confirm that my changes to the second sentence retain the intended meaning.

Comment (natebower) [inline _clients/data-prepper/index.md:38]:
```suggestion
Sink is the output component of a pipeline that defines the destination(s) to which a Data Prepper pipeline publishes events. A sink destination could be a service, such as OpenSearch or Amazon S3, or another Data Prepper pipeline. When using another Data Prepper pipeline as the sink, you can chain multiple pipelines together based on the needs of the data. Sink contains its own configuration options based on the destination type.
```

File: _clients/data-prepper/index.md [hunk 4]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+
+### Processor
+
+Processors are units within the Data Prepper pipeline that can filter, transform, and enrich events into your desired format before publishing the record to the sink. The a processor is not defined in the pipeline configuration, the events publish in the format defined in the source component. You can have more than on processor within a pipeline. When using multiple processors, the processors are executed in the order they are defined inside the pipeline spec.
```
Comment (natebower) [inline _clients/data-prepper/index.md:42]:
```suggestion
Processors are units within the Data Prepper pipeline that can filter, transform, and enrich events into your desired format before publishing the record to the sink component. The processor is not defined in the pipeline configuration; the events publish in the format defined in the source component. You can have more than one processor within a pipeline. When using multiple processors, the processors are run in the order they are defined inside the pipeline spec.
```

Comment (natebower) [inline _clients/data-prepper/index.md:42]: In the first sentence, it looks like "into" does not agree with all three verbs. Would "using" work? Last word: Can "spec" be "specification"?

File: _clients/data-prepper/index.md [hunk 5]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+
+## Sample Pipeline configurations
```
Comment (natebower) [inline _clients/data-prepper/index.md:44]: Should "Pipeline" be lowercase?

File: _clients/data-prepper/index.md [hunk 6]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+
+To understand how all pipeline components function within a Data Prepper configuration, see the following examples. Each pipeline configuration uses a `yaml` file format.
+
+### Minimal component
+
+This pipeline configuration reads from file source and writes to that same source. It uses the default options for buffer and processor.
```
Comment (natebower) [inline _clients/data-prepper/index.md:50]: Should this be "This pipeline configuration reads from the file source and writes to that same source. It uses the default options for the buffer and processor"?

File: _clients/data-prepper/index.md [hunk 7]
```diff
@@ -12,4 +12,77 @@ Data Prepper is a server side data collector capable of filtering, enriching, tr
...
+    file:
+        path: <path/to/input-file>
+  sink:
+    - file:
+        path: <path/to/output-file>
+```
+
+### All components
+
+The following pipeline uses a source that reads string events from the `input-file`. The source then pushes the data to buffer bounded by max size of `1024`. The pipeline configured to have `4` workers each of them reading maximum of `256` events from the buffer for every `100 milliseconds`. Each worker executes the `string_converter` processor and write the output of the processor to the `output-file`.
```
Comment (natebower) [inline _clients/data-prepper/index.md:64]:
```suggestion
The following pipeline uses a source that reads string events from the `input-file`. The source then pushes the data to the buffer, bounded by a max size of `1024`. The pipeline is configured to have `4` workers, each of them reading a maximum of `256` events from the buffer for every `100 milliseconds`. Each worker runs the `string_converter` processor and writes the output of the processor to the `output-file`.
```

**Review by natebower — COMMENTED** [review]
@Naarcha-AWS Please see my changes and comments, and let me know if you have any questions. Thanks!

File: _clients/data-prepper/index.md (hunk 3 above)
Comment (Naarcha-AWS) [reply _clients/data-prepper/index.md:38]: @natebower: For Source and Sink we need to indicate that they are input and output components, in order for technical users to understand the parameters they are configuring for each.

Which of these do you prefer:

A. "The sink output component defines the destination..."
B. "Sink is the output component that defines the destination(s) to which a Data Prepper pipeline publishes events."

File: _clients/data-prepper/index.md (hunk 4 above)
Comment (Naarcha-AWS) [reply _clients/data-prepper/index.md:42]: Works for me.

File: _clients/data-prepper/index.md (hunk 5 above)
Comment (Naarcha-AWS) [inline _clients/data-prepper/index.md:44]:
```suggestion
## Sample pipeline configurations
```

File: _clients/data-prepper/index.md (hunk 6 above)
Comment (Naarcha-AWS) [inline _clients/data-prepper/index.md:50]:
```suggestion
This pipeline configuration reads from the file source and writes to that same source. It uses the default options for the buffer and processor.
```

**Comment by carolxob** [comment]
@Naarcha-AWS  Did you mean to close this PR? Does it need more reviews?

**Comment by carolxob** [comment]
Overview has been published on this page: https://opensearch.org/docs/latest/data-prepper/index/

**Pull request closed without merging**